File: src/tools/edit.zig
const std = @import("std");
const core = @import("phoenix_core");

const max_bytes: usize = 512 * 1024;

fn errResult(allocator: std.mem.Allocator, comptime fmt: []const u8, args: anytype) anyerror!core.tool.ToolResult {
    const msg = try std.fmt.allocPrint(allocator, fmt, args);
    return .{ .output = msg, .is_error = true };
}

fn invoke(self: *const core.Tool, io: std.Io, args: []const u8, allocator: std.mem.Allocator) anyerror!core.tool.ToolResult {
    _ = self;

    var parsed = std.json.parseFromSlice(std.json.Value, allocator, args, .{}) catch {
        return errResult(allocator, "edit: invalid JSON args: {s}", .{args});
    };
    defer parsed.deinit();
    if (parsed.value != .object) return errResult(allocator, "edit: args must be a JSON object", .{});
    const obj = parsed.value.object;

    const path_v = obj.get("path") orelse return errResult(allocator, "edit: missing required field 'path'", .{});
    if (path_v != .string) return errResult(allocator, "edit: 'path' must be a string", .{});
    const path = path_v.string;

    const old_v = obj.get("oldText") orelse return errResult(allocator, "edit: missing required field 'oldText'", .{});
    if (old_v != .string) return errResult(allocator, "edit: 'oldText' must be a string", .{});
    const old_text = old_v.string;

    const new_v = obj.get("newText") orelse return errResult(allocator, "edit: missing required field 'newText'", .{});
    if (new_v != .string) return errResult(allocator, "edit: 'newText' must be a string", .{});
    const new_text = new_v.string;

    if (old_text.len == 0) return errResult(allocator, "edit: 'oldText' must not be empty", .{});

    const orig = std.Io.Dir.cwd().readFileAlloc(io, path, allocator, .limited(max_bytes)) catch |err| {
        return errResult(allocator, "edit: cannot read {s}: {s}", .{ path, @errorName(err) });
    };
    defer allocator.free(orig);

    const idx = std.mem.indexOf(u8, orig, old_text) orelse {
        return errResult(allocator, "edit: oldText not found in {s}", .{path});
    };

    var out: std.ArrayList(u8) = .empty;
    defer out.deinit(allocator);
    try out.appendSlice(allocator, orig[0..idx]);
    try out.appendSlice(allocator, new_text);
    try out.appendSlice(allocator, orig[idx + old_text.len ..]);

    std.Io.Dir.cwd().writeFile(io, .{ .sub_path = path, .data = out.items }) catch |err| {
        return errResult(allocator, "edit: cannot write {s}: {s}", .{ path, @errorName(err) });
    };

    const msg = try std.fmt.allocPrint(allocator, "edited {s} ({d} -> {d} bytes)", .{ path, orig.len, out.items.len });
    return .{ .output = msg };
}

pub const edit_tool: core.Tool = .{
    .name = "edit",
    .description = "Edit a file by replacing exact text. The oldText must match exactly (including whitespace). Use this for precise, surgical edits.",
    .schema =
    \\{"type":"object","properties":{"path":{"type":"string","description":"Path to the file to edit (relative or absolute)"},"oldText":{"type":"string","description":"Exact text to find and replace (must match exactly)"},"newText":{"type":"string","description":"New text to replace the old text with"}},"required":["path","oldText","newText"]}
    ,
    .max_output_bytes = max_bytes,
    .invokeFn = invoke,
};

// ---- Tests ----

test "edit replaces exact text" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const tmp_path = "phoenix_test_edit.txt";
    try std.Io.Dir.cwd().writeFile(io, .{ .sub_path = tmp_path, .data = "hello world\nfoo bar baz\n" });
    defer std.Io.Dir.cwd().deleteFile(io, tmp_path) catch {};

    const args =
        \\{"path":"phoenix_test_edit.txt","oldText":"foo bar","newText":"FOO BAR"}
    ;
    const r = try edit_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expect(!r.is_error);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "edited") != null);

    const back = try std.Io.Dir.cwd().readFileAlloc(io, tmp_path, a, .limited(128));
    defer a.free(back);
    try std.testing.expectEqualStrings("hello world\nFOO BAR baz\n", back);
}

test "edit returns error when oldText not found" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const tmp_path = "phoenix_test_edit_missing.txt";
    try std.Io.Dir.cwd().writeFile(io, .{ .sub_path = tmp_path, .data = "nothing matches here\n" });
    defer std.Io.Dir.cwd().deleteFile(io, tmp_path) catch {};

    const args =
        \\{"path":"phoenix_test_edit_missing.txt","oldText":"absent","newText":"present"}
    ;
    const r = try edit_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expect(r.is_error);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "oldText not found") != null);
}

test "edit empty oldText is error" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const args =
        \\{"path":"/tmp/x","oldText":"","newText":"y"}
    ;
    const r = try edit_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expect(r.is_error);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "oldText") != null);
}

File: src/tools/edit.rs
use async_trait::async_trait;
use serde_json::{Value, json};

use super::traits::{InputRequester, Tool, ToolError, ToolResult, ToolSchema};

pub struct EditTool;

#[async_trait]
impl Tool for EditTool {
    fn schema(&self) -> ToolSchema {
        ToolSchema {
            name: "edit".into(),
            description: "Edit a file by replacing exact text. The old_string must match exactly \
                          (including whitespace). Fails if old_string is not unique unless \
                          replace_all is true."
                .into(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "file_path": {
                        "type": "string",
                        "description": "Absolute path to the file to edit"
                    },
                    "old_string": {
                        "type": "string",
                        "description": "Exact text to find and replace (must match exactly)"
                    },
                    "new_string": {
                        "type": "string",
                        "description": "New text to replace the old text with"
                    },
                    "replace_all": {
                        "type": "boolean",
                        "description": "If true, replace all occurrences; otherwise fail on non-unique match",
                        "default": false
                    }
                },
                "required": ["file_path", "old_string", "new_string"]
            }),
        }
    }

    async fn invoke(
        &self,
        args: Value,
        _input: &dyn InputRequester,
    ) -> Result<ToolResult, ToolError> {
        let file_path = args
            .get("file_path")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::InvalidArgs("missing required field 'file_path'".into()))?;

        let old_string = args
            .get("old_string")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::InvalidArgs("missing required field 'old_string'".into()))?;

        let new_string = args
            .get("new_string")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::InvalidArgs("missing required field 'new_string'".into()))?;

        let replace_all = args
            .get("replace_all")
            .and_then(|v| v.as_bool())
            .unwrap_or(false);

        if old_string.is_empty() {
            return Err(ToolError::InvalidArgs(
                "'old_string' must not be empty".into(),
            ));
        }

        // Read the file.
        let original = tokio::fs::read_to_string(file_path)
            .await
            .map_err(|e| ToolError::ExecutionFailed(format!("cannot read {file_path}: {e}")))?;

        // Count occurrences.
        let count = original.matches(old_string).count();

        if count == 0 {
            return Err(ToolError::ExecutionFailed(format!(
                "old_string not found in {file_path}"
            )));
        }

        if count > 1 && !replace_all {
            return Err(ToolError::ExecutionFailed(format!(
                "old_string found {count} times in {file_path}; use replace_all to replace all occurrences"
            )));
        }

        // Perform the replacement.
        let updated = if replace_all {
            original.replace(old_string, new_string)
        } else {
            original.replacen(old_string, new_string, 1)
        };

        tokio::fs::write(file_path, &updated)
            .await
            .map_err(|e| ToolError::ExecutionFailed(format!("cannot write {file_path}: {e}")))?;

        let mut diff = format!("edited {file_path}: replaced {count} occurrence(s)\n");
        for line in old_string.lines() {
            diff.push_str(&format!("- {line}\n"));
        }
        for line in new_string.lines() {
            diff.push_str(&format!("+ {line}\n"));
        }
        Ok(ToolResult::success(diff))
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(all(test, not(miri)))]
mod tests {
    use super::*;
    use crate::tools::traits::NoopInputRequester;
    use tempfile::TempDir;

    fn write_temp(dir: &TempDir, name: &str, content: &str) -> String {
        let path = dir.path().join(name);
        std::fs::write(&path, content).unwrap();
        path.to_str().unwrap().to_string()
    }

    #[tokio::test]
    async fn edit_replaces_unique_string() {
        let dir = TempDir::new().unwrap();
        let path = write_temp(&dir, "test.txt", "hello world");

        let result = EditTool
            .invoke(
                json!({
                    "file_path": path,
                    "old_string": "hello",
                    "new_string": "goodbye"
                }),
                &NoopInputRequester,
            )
            .await
            .unwrap();

        assert!(!result.is_error);
        assert_eq!(std::fs::read_to_string(&path).unwrap(), "goodbye world");
    }

    #[tokio::test]
    async fn edit_fails_on_non_unique() {
        let dir = TempDir::new().unwrap();
        let path = write_temp(&dir, "dup.txt", "aaa bbb aaa");

        let result = EditTool
            .invoke(
                json!({
                    "file_path": path,
                    "old_string": "aaa",
                    "new_string": "zzz"
                }),
                &NoopInputRequester,
            )
            .await;

        match result {
            Err(ToolError::ExecutionFailed(msg)) => {
                assert!(msg.contains("2 times"), "msg was: {msg}");
                assert!(msg.contains("replace_all"), "msg was: {msg}");
            }
            other => panic!("expected ExecutionFailed, got: {other:?}"),
        }
        // File should be unchanged.
        assert_eq!(std::fs::read_to_string(&path).unwrap(), "aaa bbb aaa");
    }

    #[tokio::test]
    async fn edit_replace_all() {
        let dir = TempDir::new().unwrap();
        let path = write_temp(&dir, "all.txt", "aaa bbb aaa ccc aaa");

        let result = EditTool
            .invoke(
                json!({
                    "file_path": path,
                    "old_string": "aaa",
                    "new_string": "zzz",
                    "replace_all": true
                }),
                &NoopInputRequester,
            )
            .await
            .unwrap();

        assert!(!result.is_error);
        assert_eq!(
            std::fs::read_to_string(&path).unwrap(),
            "zzz bbb zzz ccc zzz"
        );
    }

    #[tokio::test]
    async fn edit_old_string_not_found() {
        let dir = TempDir::new().unwrap();
        let path = write_temp(&dir, "nf.txt", "nothing here");

        let result = EditTool
            .invoke(
                json!({
                    "file_path": path,
                    "old_string": "absent",
                    "new_string": "present"
                }),
                &NoopInputRequester,
            )
            .await;

        match result {
            Err(ToolError::ExecutionFailed(msg)) => {
                assert!(msg.contains("not found"), "msg was: {msg}");
            }
            other => panic!("expected ExecutionFailed, got: {other:?}"),
        }
    }

    #[tokio::test]
    async fn edit_empty_old_string() {
        let result = EditTool
            .invoke(
                json!({
                    "file_path": "/tmp/x",
                    "old_string": "",
                    "new_string": "y"
                }),
                &NoopInputRequester,
            )
            .await;

        match result {
            Err(ToolError::InvalidArgs(msg)) => {
                assert!(msg.contains("old_string"), "msg was: {msg}");
            }
            other => panic!("expected InvalidArgs, got: {other:?}"),
        }
    }

    #[tokio::test]
    async fn edit_missing_file() {
        let result = EditTool
            .invoke(
                json!({
                    "file_path": "/tmp/nonexistent_phx_edit_test.txt",
                    "old_string": "a",
                    "new_string": "b"
                }),
                &NoopInputRequester,
            )
            .await;

        match result {
            Err(ToolError::ExecutionFailed(msg)) => {
                assert!(msg.contains("cannot read"), "msg was: {msg}");
            }
            other => panic!("expected ExecutionFailed, got: {other:?}"),
        }
    }

    #[tokio::test]
    async fn edit_preserves_whitespace() {
        let dir = TempDir::new().unwrap();
        let path = write_temp(&dir, "ws.txt", "  hello  \n  world  \n");

        let result = EditTool
            .invoke(
                json!({
                    "file_path": path,
                    "old_string": "  hello  ",
                    "new_string": "  goodbye  "
                }),
                &NoopInputRequester,
            )
            .await
            .unwrap();

        assert!(!result.is_error);
        assert_eq!(
            std::fs::read_to_string(&path).unwrap(),
            "  goodbye  \n  world  \n"
        );
    }
}
